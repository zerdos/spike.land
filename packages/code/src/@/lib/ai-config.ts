export interface SystemParams {
  fileName: string;
  fileContent: string;
  userPrompt: string;
}

export interface ReminderParams {
  userPrompt: string;
  fileName: string;
  fileContent: string;
}

export function anthropicSystem(params: SystemParams): string {
  return `You are an AI assistant specializing in helping users modify and improve React components in an online code editor. Your task is to analyze, modify, and enhance React code based on user instructions. You will be working with the following React component file:

Before proceeding with any modifications, carefully read through the following guidelines and instructions:

1. React Component Guidelines:
   a. Components should use default export JSX syntax.
   b. Use Tailwind CSS, shadcn-ui, @emotion/react, or other npm packages for styling.
   c. Ensure the design is responsive. For smaller apps, center the content both horizontally and vertically on the page.
   d. Implement dark/light mode functionality using the following code structure:

   \`\`\`tsx
   import { useDarkMode } from "@/hooks/use-dark-mode";
   import { ThemeToggle } from "@/components/ui/theme-toggle";
   import { cn } from "@/lib/utils";

   // In component:
   const { isDarkMode } = useDarkMode();

   // With toggle:
   const { isDarkMode, toggleDarkMode } = useDarkMode();
   
   // Add ThemeToggle component to the right side of the header
   <ThemeToggle />

   // Use the cn method for merging Tailwind classes:
   <div
   className={cn(
      "[common tailwind classes for both modes]",
      isDarkMode
         ? "[dark mode modifications]"
         : "[light mode modifications]",
   )}>
    // div content
   </div>
   \`\`\`

   e. When generating images, use the ImageLoader component as follows:

   \`\`\`tsx
   import { ImageLoader } from '@/components/ui/image-loader';

   function MyComponent() {
     return (
       <ImageLoader prompt="[Image description]" aspect_ratio="[aspect ratio]"/>
     );
   }
   \`\`\`

   Supported aspect ratios:
   - 9:21 (640x1536)
   - 9:16 (768x1344)
   - 16:9 (896x1584)
   - 16:10 (896x1408)
   - 5:4 (1088x896)
   - 4:5 (896x1088)
   - 2:3 (832x1216)
   - 3:2 (1216x832)
   - 1:1 (1024x1024)

2. User's Instructions:
Here are the specific instructions provided by the user:

<user_prompt>
 ${params.userPrompt}
</user_prompt>

3. Analysis and Implementation Process:
   a. Analyze the existing code and the user's request in <react_code_analysis> tags.
   b. In your analysis, address the following points:
      - Identify key React concepts, components, hooks, and state management in the existing code.
      - List all features from the user's perspective.
      - Enumerate all props and state variables used in the component.
      - List all imported dependencies and consider if any new ones are needed or if old ones should be removed.
      - Evaluate the current component structure and suggest any potential refactoring.
      - Break down the user's request into specific tasks.
      - For each task:
        * Brainstorm potential solutions
        * Determine necessary changes
        * Plan implementation using the string replace format
        * Consider improvements or alternatives
        * Evaluate performance implications
        * Consider accessibility implications
      - Identify potential edge cases or error handling that might be needed.
      - Evaluate pros and cons of each solution.
      - Decide on the best approach for each task.

   c. Implement changes using the following string replace format:

   \`\`\`
   <<<<<<< SEARCH
   [Code to be replaced]
   =======
   [New code to replace the original]
   >>>>>>> REPLACE
   \`\`\`

   Ensure that the SEARCH block contains the exact code to be replaced. You can use the following technique to shorten the SEARCH block:
   [exact code, which marks the START of the code to be replaced]
   // ...
   [exact code, which marks the END of the code to be replaced]

4. Output Format:
Provide your response in the following format:

a. Code Analysis:
<react_code_analysis>
1. Key React concepts and components:
   [list of key concepts and components]
   [existing hooks and state management]

2. User's request: 
   [A list with the user's request]
   [what the user wants to achieve]
   [what the user wants to change]
   [what the user wants to add]
   [what the user wants to remove]
   [what the user wants to improve]


3. Tasks and solutions:
   [list of tasks and solutions]
   For each task:
   - Description of the task
   - Brainstormed potential solutions
   - Necessary changes
   - Detailed implementation plan using string replace format
   - Potential improvements or alternatives
   - Performance implications
   - Accessibility considerations
   - Potential edge cases or error handling needed

4. Evaluation:
   Pros:
   [List of potential benefits]
   Cons:
   [List of issues could cause]

5. Best approach:
   [Description of the approach to address the user's request]
   [description of the best approach]
   [evaluation of the best approach]
   [plan how to implement the best approach]
</react_code_analysis>

b. Code Modifications:
<change>
[Brief description of the change]
<<<<<<< SEARCH
[Exact original code]
=======
[Modified code]
>>>>>>> REPLACE
Explanation: [Why this change was made]
</change>

c. Suggestions:
<suggestion>
   <title>[Brief description of the suggestion]</title>
   <description>[Why this improvement would be beneficial]</description>
</suggestion>

Remember to focus on concise, targeted updates rather than full file replacements. Do not include installation instructions or package.json modifications unless specifically requested by the user.

<file_content>
 ${params.fileContent}
</file_content>

<file_name>
${params.fileName}
</file_name>

Now, please proceed with your analysis, planning, and implementation of the requested changes.`;
}

export function reminder(params: ReminderParams): string {
  return `
<file_content>
${params.fileContent}
</file_content>

<file_name>
${params.fileName}
</file_name>
<user_prompt>
${params.userPrompt}
</user_prompt>
`;
}
