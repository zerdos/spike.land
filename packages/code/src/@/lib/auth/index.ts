/**
 * Unified Authentication Library
 * Provides consistent authentication across all spike.land packages
 *
 * Security Features:
 * - Multiple auth provider support (Clerk, JWT, OAuth)
 * - OWASP compliant security practices
 * - Rate limiting and CSRF protection
 * - Session management with automatic refresh
 * - Audit logging for security events
 * - MFA support
 * - IP whitelisting/blacklisting
 * - Secure password policies
 */

// Export types
export * from "./types";

// Export auth manager
export { AuthManager, createAuthManager } from "./manager";

// Export React hooks
export {
  useAuth,
  useUser,
  useSession,
  useRequireAuth,
  usePermissions,
  useAuthError,
  useMFA,
  useSessionActivity,
  useAuthProvider,
} from "./hooks";

// Export middleware
export {
  createAuthMiddleware,
  createCSRFMiddleware,
  createSessionMiddleware,
  createSecurityHeadersMiddleware,
  createAPIKeyMiddleware,
  createWebhookMiddleware,
  authMiddleware,
  type AuthMiddlewareOptions,
} from "./middleware";

// Export providers
export { BaseAuthProvider } from "./providers/base";
export { ClerkAuthProvider } from "./providers/clerk";
export { JWTAuthProvider } from "./providers/jwt";

// Export security utilities
export { securityUtils } from "./security";

// Default export for convenience
import { createAuthManager } from "./manager";
import { authMiddleware } from "./middleware";

export default {
  createManager: createAuthManager,
  middleware: authMiddleware,
};