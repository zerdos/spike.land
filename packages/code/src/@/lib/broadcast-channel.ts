const broadcastChannels = new Map<string, BroadcastChannel>();

export const getBroadcastChannel = (codeSpace: string) => {
  if (broadcastChannels.has(codeSpace)) {
    return broadcastChannels.get(codeSpace)!;
  }

  const bc = new BroadcastChannel(`codeSpace-${codeSpace}`);

  broadcastChannels.set(codeSpace, bc);

  return bc;
};
