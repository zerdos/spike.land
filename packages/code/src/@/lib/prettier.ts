import emotionPrettifier from "@emotion/css-prettifier";
import type { Options } from "prettier";
import pluginEstree from "prettier/plugins/estree";
import postCss from "prettier/plugins/postcss";
import pluginTypescript from "prettier/plugins/typescript";
import { format } from "prettier/standalone";
import { tryCatch } from "./try-catch"; // Added import

/* ============================================================================
   Utility Functions
   ========================================================================== */

/**
 * Generates a string of spaces for a given length.
 * @param length - The number of spaces.
 * @returns A string of spaces.
 */
const createSpaceString = (length: number): string => " ".repeat(length);

/**
 * Removes block comments matching the pattern {/** ...
 * @param code - The code to process.
 * @returns The code without the block comments.
 */
const removeBlockComments = (code: string): string => {
  const blockCommentRegex = /\{\/\*\*([\s\S]*?)\*\/\}/g;
  return code.replace(blockCommentRegex, "");
};

/* ============================================================================
   CSS-in-JS Processing Functions
   ========================================================================== */

/**
 * Formats CSS content within a template literal.
 * It replaces interpolations with placeholders, formats the CSS,
 * then restores the interpolations.
 *
 * @param cssContent - The raw CSS content.
 * @param indent - The indentation string.
 * @returns The formatted CSS content with interpolations restored.
 */
const formatCssContent = (cssContent: string, indent: string): string => {
  const placeholderPrefix = "__INTERPOLATION_PLACEHOLDER_";
  const interpolations: string[] = [];
  let placeholderIndex = 0;

  // Replace interpolations with unique placeholders.
  const contentWithPlaceholders = cssContent.replace(/\${(.*?)}/g, (match) => {
    const placeholder = `${placeholderPrefix}${placeholderIndex}`;
    interpolations.push(match);
    placeholderIndex++;
    return placeholder;
  });

  // Format the CSS content using the emotion prettifier.
  const formattedContent = emotionPrettifier(contentWithPlaceholders)
    .split("\n")
    .map((line) => (line.trim() ? `${indent}  ${line}` : line))
    .join("\n");

  // Restore the interpolations.
  return formattedContent.replace(
    new RegExp(`${placeholderPrefix}\\d+`, "g"),
    (match) => {
      const index = parseInt(match.replace(placeholderPrefix, ""), 10);
      return interpolations[index] || "";
    },
  );
};

/**
 * Adds a default export if none is found in the code.
 *
 * @param code - The code to check.
 * @returns The code with a default export appended if necessary.
 */
const addDefaultExportIfNeeded = (code: string): string => {
  if (
    !code.includes("export default") &&
    !code.includes("export const") &&
    !code.includes("const App")
  ) {
    return `${code}\n\nexport default () => <></>; // Empty default export`;
  }
  return code;
};

/**
 * Processes and formats CSS-in-JS code that uses the css template literal.
 * It also inserts an Emotion import if not present.
 *
 * @param rawCode - The original code string.
 * @returns The processed and formatted code.
 */
export const addSomeFixesIfNeeded = (rawCode: string): string => {
  const code = removeBlockComments(rawCode);
  try {
    const parts = code.split("css`");
    let header = parts.shift() || "";

    // Insert Emotion's css import if it's missing.
    if (
      parts.length && !code.includes("@emotion/react") &&
      !code.includes(" css ")
    ) {
      const [firstLine, ...restLines] = header.split("\n");
      header = (firstLine || "").startsWith("//")
        ? [firstLine, 'import { css } from "@emotion/react";', ...restLines].join(
          "\n",
        )
        : ['import { css } from "@emotion/react";', firstLine, ...restLines]
          .join("\n");
    }

    let currentIndent = (header.split("\n").pop()?.length || 0) + 2;

    // Process each css template literal part.
    const processedSections = parts.map((section) => {
      const [cssContent, afterCss, ...remainingParts] = section.split("`");
      const indent = createSpaceString(currentIndent);
      currentIndent = (afterCss?.split("\n").pop()?.length || 0) + 2;

      const formattedCss = formatCssContent(cssContent || "", indent);
      return `${formattedCss}\n${indent}\`${[afterCss, ...remainingParts].join("`")}`;
    });

    let result = [header, ...processedSections].join("css`\n");
    result = addDefaultExportIfNeeded(result);
    return result;
  } catch (error) {
    console.error("addSomeFixesIfNeeded error", { error, code });
    return code;
  }
};

/* ============================================================================
   Prettier Formatting Functions & Configuration
   ========================================================================== */

/**
 * Prettier configuration for JavaScript/TypeScript.
 */
const prettierConfig: Options = {
  arrowParens: "always",
  bracketSpacing: true,
  embeddedLanguageFormatting: "off",
  insertPragma: false,
  bracketSameLine: false,
  jsxSingleQuote: false,
  htmlWhitespaceSensitivity: "strict",
  printWidth: 80,
  proseWrap: "preserve",
  quoteProps: "as-needed",
  requirePragma: false,
  semi: true,
  singleQuote: false,
  tabWidth: 2,
  trailingComma: "es5",
  useTabs: false,
  parser: "typescript",
  singleAttributePerLine: false,
  plugins: [pluginEstree, pluginTypescript],
};

/**
 * Formats JavaScript/TypeScript code using Prettier.
 *
 * @param params.code - The code to format.
 * @param params.toThrow - If true, errors are thrown.
 * @returns The formatted code.
 */
export const prettierJs = async ({
  code,
  toThrow,
}: {
  code: string;
  toThrow: boolean;
}): Promise<string> => {
  const formatPromise = format(code, prettierConfig);
  const { data, error } = await tryCatch(formatPromise);

  if (error) {
    if (toThrow) throw error;
    console.warn(
      "Prettier JS formatting failed, returning original code:",
      error,
    );
    return code;
  }
  return data || code; // Return formatted code or original if data is null/undefined (should not happen with format)
};

/**
 * Formats CSS code using Prettier with a CSS parser.
 *
 * @param inputCSS - The CSS code to format.
 * @returns The formatted CSS.
 */
export const prettierCss = async (inputCSS: string): Promise<string> => {
  const formatPromise = format(inputCSS, {
    parser: "css",
    printWidth: 80,
    tabWidth: 2,
    useTabs: false,
    singleQuote: false,
    trailingComma: "none",
    plugins: [postCss],
  });
  const { data, error } = await tryCatch(formatPromise);

  if (error) {
    console.warn(
      "Prettier CSS formatting failed, returning original CSS:",
      error,
    );
    return inputCSS;
  }
  return data || inputCSS; // Return formatted CSS or original if data is null/undefined
};
