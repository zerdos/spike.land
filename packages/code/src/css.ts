/**
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import type { AnyNode, ChildNode, Rule } from "postcss";
import { parse, stringify } from "postcss";
import type Root_ from "postcss/lib/root";

/**
 * Parse a textual CSS Stylesheet into a Stylesheet instance.
 * Stylesheet is a mutable postcss AST with format similar to CSSOM.
 * @see https://github.com/postcss/postcss/
 * @private
 */
export function parseStylesheet(stylesheet: string) {
  return parse(stylesheet);
}

/**
 * Options used by the stringify logic
 */
interface SerializeStylesheetOptions {
  /** Compress CSS output (removes comments, whitespace, etc) */
  compress?: boolean;
}

/**
 * Serialize a postcss Stylesheet to a String of CSS.
 * @private
 * @param ast A Stylesheet to serialize, such as one returned from `parseStylesheet()`
 */
export function serializeStylesheet(
  ast: AnyNode,
  options: SerializeStylesheetOptions,
) {
  const cssParts: string[] = [];

  stringify(ast, (result, node, type) => {
    if (node?.type === "decl" && node.value.includes("</style>")) {
      return;
    }

    if (!options.compress) {
      cssParts.push(result);
      return;
    }

    // Simple minification logic
    if (node?.type === "comment") {
      return;
    }

    if (node?.type === "decl") {
      const prefix = node.prop + node.raws.between;

      cssParts.push(result.replace(prefix, prefix.trim()));
      return;
    }

    if (type === "start") {
      if (node?.type === "rule" && node.selectors) {
        if (node.selectors.length === 1) {
          cssParts.push(node.selectors[0] ?? "", "{");
        } else {
          cssParts.push(node.selectors.join(","), "{");
        }
      } else {
        cssParts.push(result.trim());
      }
      return;
    }

    if (type === "end" && result === "}" && node?.raws?.semicolon) {
      const lastItemIdx = cssParts.length - 2;
      if (lastItemIdx >= 0 && cssParts[lastItemIdx]) {
        cssParts[lastItemIdx] = cssParts[lastItemIdx].slice(0, -1);
      }
    }

    cssParts.push(result.trim());
  });

  return cssParts.join("");
}

type SingleIterator<T> = (item: T) => boolean | void;

/**
 * Converts a walkStyleRules() iterator to mark nodes with `.$$remove=true` instead of actually removing them.
 * This means they can be removed in a second pass, allowing the first pass to be nondestructive (eg: to preserve mirrored sheets).
 * @private
 * @param predicate   Invoked on each node in the tree. Return `false` to remove that node.
 */
export function markOnly(
  predicate: SingleIterator<ChildNode | Root_>,
): (rule: Rule | ChildNode | Root_) => void {
  return (rule) => {
    const sel = "selectors" in rule ? rule.selectors : undefined;
    if (predicate(rule) === false) {
      rule.$$remove = true;
    }
    if ("selectors" in rule) {
      rule.$$markedSelectors = rule.selectors;
      rule.selectors = sel!;
    }
    if (rule._other) {
      rule._other.$$markedSelectors = rule._other.selectors;
    }
  };
}

/**
 * Apply filtered selectors to a rule from a previous markOnly run.
 * @private
 * @param rule The Rule to apply marked selectors to (if they exist).
 */
export function applyMarkedSelectors(rule: Rule) {
  if (rule.$$markedSelectors) {
    rule.selectors = rule.$$markedSelectors;
  }
  if (rule._other) {
    applyMarkedSelectors(rule._other);
  }
}

/**
 * Recursively walk all rules in a stylesheet.
 * @private
 * @param node       A Stylesheet or Rule to descend into.
 * @param iterator   Invoked on each node in the tree. Return `false` to remove that node.
 */
export function walkStyleRules(
  node: ChildNode | Root_,
  iterator: SingleIterator<ChildNode | Root_ | Rule>,
) {
  if (!("nodes" in node)) {
    return;
  }
  node.nodes = node.nodes?.filter((rule) => {
    if (hasNestedRules(rule)) {
      walkStyleRules(rule, iterator);
    }
    rule.filterSelectors = filterSelectors;
    return iterator(rule) !== false;
  });
}

/**
 * Recursively walk all rules in two identical stylesheets, filtering nodes into one or the other based on a predicate.
 * @private
 * @param node       A Stylesheet or Rule to descend into.
 * @param node2      A second tree identical to `node`
 * @param iterator   Invoked on each node in the tree. Return `false` to remove that node from the first tree, true to remove it from the second.
 */
export function walkStyleRulesWithReverseMirror(
  node: Rule | Root_,
  node2: Rule | Root_ | undefined | null,
  iterator: SingleIterator<ChildNode | Root_>,
) {
  if (!node2) {
    return walkStyleRules(node, iterator);
  }

  [node.nodes, node2.nodes] = splitFilter(
    node.nodes,
    node2.nodes,
    (rule, index, _rules, rules2) => {
      const rule2 = rules2?.[index];
      if (hasNestedRules(rule)) {
        walkStyleRulesWithReverseMirror(rule, rule2 as Rule, iterator);
      }
      rule._other = rule2 as Rule;
      rule.filterSelectors = filterSelectors;
      return iterator(rule) !== false;
    },
  );
}

// Checks if a node has nested rules, like @media
// @keyframes are an exception since they are evaluated as a whole
function hasNestedRules(rule: ChildNode): rule is Rule {
  return (
    "nodes" in rule &&
    !!rule.nodes?.length &&
    (!("name" in rule) ||
      (rule.name !== "keyframes" && rule.name !== "-webkit-keyframes")) &&
    rule.nodes.some((n) => n.type === "rule" || n.type === "atrule")
  );
}

// Like [].filter(), but applies the opposite filtering result to a second copy of the Array without a second pass.
// This is just a quicker version of generating the compliment of the set returned from a filter operation.
type SplitIterator<T> = (item: T, index: number, a: T[], b?: T[]) => boolean;
function splitFilter<T>(a: T[], b: T[], predicate: SplitIterator<T>) {
  const aOut: T[] = [];
  const bOut: T[] = [];
  for (let index = 0; index < a.length; index++) {
    const item = a[index]!;
    if (predicate(item, index, a, b)) {
      aOut.push(item);
    } else {
      bOut.push(item);
    }
  }
  return [aOut, bOut] as const;
}

// can be invoked on a style rule to subset its selectors (with reverse mirroring)
function filterSelectors(this: Rule, predicate: SplitIterator<string>) {
  if (this._other) {
    const [a, b] = splitFilter(
      this.selectors,
      this._other.selectors,
      predicate,
    );
    this.selectors = a;
    this._other.selectors = b;
  } else {
    this.selectors = this.selectors.filter(predicate);
  }
}

declare module "postcss" {
  interface Node {
    _other?: Rule;
    $$remove?: boolean;
    $$markedSelectors?: string[];
    filterSelectors?: typeof filterSelectors;
  }
}
