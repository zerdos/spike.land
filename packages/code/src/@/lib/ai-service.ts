import { anthropicSystem, reminder } from "@/lib/ai-config";
import type { Message } from "@/lib/interfaces";
import Anthropic from "@anthropic-ai/sdk";
import OpenAI from "openai";

export interface AIServiceConfig {
  anthropicEndpoint: string;
  openAIEndpoint: string;
  gpt4oEndpoint: string;
  updateThrottleMs: number;
  retryWithClaudeEnabled: boolean;
  setIsStreaming: (isStreaming: boolean) => void;
}

export class AIService {
  private _config: AIServiceConfig;

  constructor(config: AIServiceConfig) {
    this._config = config;
  }

  public async sendToAnthropic(
    messages: Message[],
    onUpdate: (code: string) => void,
  ): Promise<Message> {
    this._config.setIsStreaming(true);

    try {
      const client = new Anthropic({
        dangerouslyAllowBrowser: true,
        apiKey: 'DUMMY_API_KEY',
        baseURL: this._config.anthropicEndpoint,
      });

      const stream = await client.messages.create({
        max_tokens: 8192,
        messages: messages.map((message) => ({
          role: message.role === "system" ? "user" : message.role,
          content: message.content.toString(),
        })),
        model: "claude-3-5-sonnet-latest",
        stream: true,
      });

      let content = "";
      for await (const messageStreamEvent of stream) {
        if (
          "type" in messageStreamEvent &&
          messageStreamEvent.type === "content_block_delta" &&
          messageStreamEvent.delta.type === "text_delta"
        ) {
          const chunkContent = messageStreamEvent.delta.text;
          if (chunkContent) {
            content += chunkContent;
            onUpdate(chunkContent);
          }
       
        }
      }

      return {
        id: Math.random().toString(36).substring(7),
        role: "assistant",
        content,
      };
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error("Unknown error occurred");
    } finally {
      this._config.setIsStreaming(false);
    }
  }

  public async sendToGpt4o(
    messages: Message[],
    onUpdate: (code: string) => void,
  ): Promise<Message> {
    this._config.setIsStreaming(true);

    try {
      const client = new OpenAI({
        baseURL: this._config.gpt4oEndpoint,
        apiKey: 'DUMMY_API_KEY',
      });

      const stream = await client.chat.completions.create({
        model: "gpt-4o-mini",
        stream: true,
        messages: messages.map((message) => ({
          role: message.role === "system" ? "user" : message.role,
          content: message.content.toString(),
        })),
      });

      let content = "";
      for await (const chunk of stream) {
        const chunkContent = chunk.choices[0]?.delta?.content;
        if (chunkContent) {
          content += chunkContent;
          onUpdate(chunkContent);
        }
      }

      return {
        id: Math.random().toString(36).substring(7),
        role: "assistant",
        content,
      };
    } catch (error) {
      if (error instanceof Error) {
        throw error;
      }
      throw new Error("Unknown error occurred");
    } finally {
      this._config.setIsStreaming(false);
    }
  }

  public prepareClaudeContent(
    params: { content: string; messages: Message[]; codeNow: string; codeSpace: string; },
  ): string {
    const { content, messages, codeNow, codeSpace } = params;

    // Check for error messages that should be passed through directly
    if (content.toLowerCase().startsWith("i'm sorry")) {
      return content;
    }

    const lastMessage = messages[messages.length - 1];

    // If codeNow is different from the last message or there are no messages,
    // use anthropic system content
    // if (!lastMessage || lastMessage.content !== codeNow) {
      const result = anthropicSystem({
        fileName: codeSpace+".tsx",
        fileContent: codeNow,
        userPrompt: content,
      });
      return result;
    // }

    // If code hasn't changed, return reminder content
    // return reminder({
    //   userPrompt: content,
    //   fileName: codeSpace+ ".tsx",
    //   fileContent: codeNow,
    // });
  }
}
