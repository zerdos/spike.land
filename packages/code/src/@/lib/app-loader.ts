import { getCodeSpace } from "@/hooks/use-code-space";
import { getCodeSession } from "@/lib/code-session";
import type { ICode } from "@/lib/interfaces";
import { init } from "@/lib/tw-dev-setup";
import { SessionSynchronizer } from "@/services/SessionSynchronizer";

// Centralized type definitions
export interface AppContext {
  codeSpace: string;
  cSess: ICode;
  AppComponent: React.FC<AppComponentProps>;
}

export interface AppComponentProps {
  codeSpace: string;
  cSess: ICode;
}

/**
 * Initializes the application environment
 * Loads modules that should be available globally
 */
export const initializeAppEnvironment = async (): Promise<void> => {
  try {
    const [
      { enhancedFetch },
      { useArchive, useSpeedy },
    ] = await Promise.all([
      import("@/lib/enhanced-fetch"),
      import("@/lib/use-archive"),
    ]);

    // Make utilities available globally for debugging and development
    Object.assign(globalThis, {
      enhancedFetch,
      useArchive,
      useSpeedy,
    });

    console.log("App environment initialized successfully");
  } catch (error) {
    console.error("Error initializing app environment:", error);
    throw error;
  }
};

/**
 * Initializes a WebSocket connection for the specified code space
 */
export const initializeWebSocket = async (codeSpace: string): Promise<void> => {
  if (!codeSpace) {
    throw new Error("Cannot initialize WebSocket: codeSpace is null or undefined");
  }

  try {
    // Import using a relative path to match the project structure
    const { main } = await import("../../ws");
    await main(codeSpace);
    console.log(`WebSocket initialized for code space: ${codeSpace}`);
  } catch (error) {
    console.error("WebSocket initialization failed:", error);
    throw error;
  }
};

/**
 * Loads the application for a given code space
 * This handles initialization of all required components and sessions
 */
export const loadApp = async (pathname: string): Promise<AppContext | null> => {
  const codeSpace = getCodeSpace(pathname);

  if (!codeSpace) {
    console.error("Cannot load app: Invalid codeSpace from pathname", pathname);
    return null;
  }

  try {
    // Initialize tailwind
    await init();

    // Get code session
    const cSess = await getCodeSession(codeSpace);

    // Load the app component dynamically
    // Import using a relative path to match the project structure
    const { AppToRender } = await import("../../AppToRender");

    // Initialize app environment
    await initializeAppEnvironment();

    return {
      codeSpace,
      cSess,
      AppComponent: AppToRender,
    };
  } catch (error) {
    console.error("Error loading app:", error);
    return null;
  }
};

/**
 * Initializes a session synchronizer for the given code space and session
 */
export const initializeSessionSync = async (
  codeSpace: string,
  cSess: ICode,
): Promise<() => void> => {
  if (!codeSpace || !cSess) {
    console.error("Cannot initialize session sync: missing codeSpace or cSess");
    return () => {};
  }

  try {
    // Make cSess available globally for debugging
    Object.assign(globalThis, { cSess });

    // Initialize session synchronizer
    const sessionSync = new SessionSynchronizer(codeSpace);
    await sessionSync.init(await cSess.getSession());

    // Return the unsubscribe function
    return sessionSync.subscribe((sess) => {
      cSess.setSession(sess);
      // Note: setState callback would be handled by the component
    });
  } catch (error) {
    console.error("Error initializing session sync:", error);
    return () => {};
  }
};

/**
 * Utility to check if a route should initialize WebSocket
 */
export const shouldInitWebSocket = (pathname: string, codeSpace: string | null): boolean => {
  if (!pathname || !codeSpace) return false;
  return pathname === `/live/${codeSpace}/iframe`;
};
