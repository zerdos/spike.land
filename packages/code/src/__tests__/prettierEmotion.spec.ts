import { addSomeFixesIfNeeded } from "@/lib/prettier";
import { describe, expect, it } from "vitest";

describe("addSomeFixesIfNeeded", () => {
  it("adds default export if 'export default' is missing and 'const App' is present", () => {
    const code = "const App = () => {};";
    expect(addSomeFixesIfNeeded(code)).toMatchSnapshot();
  });

  it("adds import statement for '@emotion/react' if 'css={css`' is present and '@emotion/react' is missing", () => {
    const code = "css={css`color: red;`}";

    expect(addSomeFixesIfNeeded(code)).toMatchSnapshot();
  });

  it("doesn't add extra css' if 'css={css`' is present and '@emotion/react' is missing", () => {
    const code = `
    import { css } from "@emotion/css";


    css={css\`color: red;\`}
    
    `;

    const result = addSomeFixesIfNeeded(code);
    expect(result.includes("@emotion/react")).toBe(false);
    expect(result).toMatchSnapshot();
  });

  it("adds import statement for 'FC' if ' FC<{' is present and 'import type { FC }' is missing", () => {
    const code = "FC<{ prop: string }>";
    expect(addSomeFixesIfNeeded(code)).toMatchSnapshot();
  });

  it("adds import statement for 'useState' if 'useState(' is present and 'import { useState }' is missing", () => {
    const code = "useState(0)";
    expect(addSomeFixesIfNeeded(code)).toMatchSnapshot();
  });

  it("handles multiple occurrences of 'css={css`' correctly", () => {
    const code = "css={css`color: red;`} css={css`font-size: 16px;`}";
    expect(addSomeFixesIfNeeded(code)).toMatchSnapshot();
  });

  it("adds default export if 'export default' is missing", () => {
    const code = "export const myUtilityFN = () => {};";
    expect(addSomeFixesIfNeeded(code)).toMatchSnapshot();
  });

  it("removes block comments", () => {
    const code = `
    <div>
    {/** This is a comment */}
    </div>`;
    expect(addSomeFixesIfNeeded(code)).toMatchSnapshot();
  });

  it("handles external state", () => {
    const code = `
    
    import { css } from '@emotion/css'

const color = 'white'

render(
  <div
    className={css\`
  padding: 32px;
          background-color: hotpink;
          font-size: 24px;
      border-radius: 4px;
     &:hover {
        color: \${color
        };
      }
    \`}
  >
    Hover to change color.
  </div>
)
  
`;
    expect(addSomeFixesIfNeeded(code)).toMatchSnapshot();
  });
});
