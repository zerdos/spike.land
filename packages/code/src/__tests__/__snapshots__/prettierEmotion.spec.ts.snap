// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addSomeFixesIfNeeded > adds default export if 'export default' is missing 1`] = `"export const myUtilityFN = () => {};"`;

exports[`addSomeFixesIfNeeded > adds default export if 'export default' is missing and 'const App' is present 1`] = `"const App = () => {};"`;

exports[`addSomeFixesIfNeeded > adds import statement for '@emotion/react' if 'css={css\`' is present and '@emotion/react' is missing 1`] = `
"import { css } from "@emotion/react";
css={css\`
         color: red;
       \`}

export default () => <></>; // Empty default export"
`;

exports[`addSomeFixesIfNeeded > adds import statement for 'FC' if ' FC<{' is present and 'import type { FC }' is missing 1`] = `
"FC<{ prop: string }>

export default () => <></>; // Empty default export"
`;

exports[`addSomeFixesIfNeeded > adds import statement for 'useState' if 'useState(' is present and 'import { useState }' is missing 1`] = `
"useState(0)

export default () => <></>; // Empty default export"
`;

exports[`addSomeFixesIfNeeded > doesn't add extra css' if 'css={css\`' is present and '@emotion/react' is missing 1`] = `
"
    import { css } from "@emotion/css";


    css={css\`
             color: red;
           \`}
    
    

export default () => <></>; // Empty default export"
`;

exports[`addSomeFixesIfNeeded > handles external state 1`] = `
"
    
    import { css } from '@emotion/css'

const color = 'white'

render(
  <div
    className={css\`
                   padding: 32px;
                   background-color: hotpink;
                   font-size: 24px;
                   border-radius: 4px;
                 \`}
  >
    Hover to change color.
  </div>
)
  


export default () => <></>; // Empty default export"
`;

exports[`addSomeFixesIfNeeded > handles multiple occurrences of 'css={css\`' correctly 1`] = `
"import { css } from "@emotion/react";
css={css\`
         color: red;
       \`} css={css\`
           font-size: 16px;
         \`}

export default () => <></>; // Empty default export"
`;

exports[`addSomeFixesIfNeeded > removes block comments 1`] = `
"
    <div>
    
    </div>

export default () => <></>; // Empty default export"
`;
